File: solutions/code-carl/ch07_backtracking/t1_lc0077_combine.py
"""LeetCode 77. Combinations
https://leetcode.cn/problems/combinations/
"""
from typing import List


class Solution:
    def combine(self, n: int, k: int) -> List[List[int]]:

        path: List[int] = []
        ans: List[List[int]] = []

        def backtracking(n: int, k: int, start_idx: int) -> None:
            if k == len(path):
                ans.append(path[:])
                return

            for i in range(start_idx, n - k + len(path) + 2):
                path.append(i)
                backtracking(n, k, i + 1)
                path.pop()

        backtracking(n, k, 1)
        return ans


if __name__ == "__main__":
    solution = Solution()
    result = {tuple(x) for x in solution.combine(4, 2)}
    expected = {(1, 2), (1, 3), (1, 4), (2, 3), (2, 4), (3, 4)}
    assert result == expected

    solution = Solution()
    assert solution.combine(1, 1) == [[1]]

    print("All tests passed.")

File: solutions/java/code-carl/ch07_backtracking/T1_LC0077_Combine.java
/**
 * Practice template for:
 * - LeetCode 77. Combinations
 *   https://leetcode.cn/problems/combinations/
 */
import java.util.ArrayList;
import java.util.List;

public class T1_LC0077_Combine {
    List<Integer> path = new ArrayList<>();
    List<List<Integer>> res = new ArrayList<>();

    public List<List<Integer>> combine(int n, int k) {
        backtracking(n, k, 1);
        return res;
    }

    public void backtracking(int n, int k, int startIdx) {
        if (path.size() == k) {
            res.add(new ArrayList<>(path));
            return;
        }
        for (int i = startIdx; i <= n - (k - path.size()) + 1; i++) {
            path.add(i);
            backtracking(n, k, i + 1);
            path.remove(path.size() - 1);
        }
    }
}
